pkg/github/github.go

```diff
@@ -0,0 +1,119 @@
+package github
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"sort"
+	"strings"
+
+	"github.com/aquasecurity/trivy-db/pkg/db"
+	"github.com/aquasecurity/trivy/pkg/log"
+	"github.com/google/go-github/v28/github"
+	"golang.org/x/oauth2"
+	"golang.org/x/xerrors"
+)
+
+const (
+	owner = "aquasecurity"
+	repo  = "trivy-db"
+)
+
+type RepositoryInterface interface {
+	ListReleases(ctx context.Context, opt *github.ListOptions) ([]*github.RepositoryRelease, *github.Response, error)
+	DownloadAsset(ctx context.Context, id int64) (io.ReadCloser, string, error)
+}
+
+type Repository struct {
+	repository *github.RepositoriesService
+	git        *github.GitService
+	owner      string
+	repoName   string
+}
+
+func (r Repository) ListReleases(ctx context.Context, opt *github.ListOptions) ([]*github.RepositoryRelease, *github.Response, error) {
+	return r.repository.ListReleases(ctx, r.owner, r.repoName, opt)
+}
+
+func (r Repository) DownloadAsset(ctx context.Context, id int64) (io.ReadCloser, string, error) {
+	return r.repository.DownloadReleaseAsset(ctx, r.owner, r.repoName, id)
+}
+
+type Client struct {
+	Repository RepositoryInterface
+}
+
+func NewClient(ctx context.Context) Client {
+	ts := oauth2.StaticTokenSource(
+		&oauth2.Token{AccessToken: os.Getenv("GITHUB_TOKEN")},
+	)
+	tc := oauth2.NewClient(ctx, ts)
+	gc := github.NewClient(tc)
+
+	repo := Repository{
+		repository: gc.Repositories,
+		git:        gc.Git,
+		owner:      owner,
+		repoName:   repo,
+	}
+
+	return Client{
+		Repository: repo,
+	}
+}
+
+func (c Client) DownloadDB(ctx context.Context, fileName string) (io.ReadCloser, error) {
+	options := github.ListOptions{}
+	releases, _, err := c.Repository.ListReleases(ctx, &options)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to list releases: %w", err)
+	}
+
+	sort.Slice(releases, func(i, j int) bool {
```

Comment:
This will panic if releases is not empty but doesn't have all the elements inside of it, in particular `PublishedAt`. For example:
```
{
  ID:   github.Int64(1),
  Name: github.String("xyzfoo1-2020123123"),
}
```

Should we explicitly check for that before proceeding to be safe?

Reply:
Great point. We should have used `GetPublishedAt`. This is safety for nil.

Reply:
Fixed.

Reply:
Fixed